Added structs and functions necessary for GPU inferencing support

lightningpwr28 opened the pull request:
I basically just made some classes in the style of the stuff that was always there, with the exception of `gpu_init()` and `gpu_thread_init()`, which I put in the root of the `vosk` crate.

I know that the recommended dynamic library binaries don't support it, but I need the support for the back end of a project I'm working on.

I think it should be possible to get binaries that do by following the compilation steps of or copying the compiled result out of the GPU containers in [vosk-server](https://github.com/alphacep/vosk-server/), at least for Ubuntu based distros. I'm working on a Windows binary, but I'm not entirely sure I'll be able to do it.

Bear-03 commented:
Wonderful work! Just some days ago I was thinking of adding exactly this, so perfect timing! In the first version of the crate I didn't wrap GPU-related code because I don't have an NVIDIA GPU myself.

> I think it should be possible to get binaries that do by following the compilation steps of or copying the compiled result out of the GPU containers in [vosk-server](https://github.com/alphacep/vosk-server/), at least for Ubuntu based distros. I'm working on a Windows binary, but I'm not entirely sure I'll be able to do it.

Yeah, as mentioned in [this issue](https://github.com/alphacep/vosk-api/issues/1586) you have to build from source with `HAVE_CUDA=1` set. Is [this](https://github.com/alphacep/vosk-api/issues/1586#issuecomment-2154617587) the Dockerfile you mention?

Also, since the gpu code is only available if you compiled your libs with the HAVE_CUDA flag, it is misleading if the rust code related to cuda is always available. I suggest putting it behind a crate feature flag, probably called `cuda`.

As to the specifics of the code, I'll review it more thoroughly as soon as possible :D

lightningpwr28 commented:
Yes, that is the one! I'll look into the feature flag and probably get that added today.

Bear-03 reviewed (changes requested):
Sorry for the delay! I've been really busy these last weeks. Anyway, I only added some minor comments of some things that probably slipped your mind, totally fair.

Another thing I wanna comment on is that I couldn't help but notice the amount of times that `#[cfg(feature = "cuda")]` is repeated. I think it'd be best to group things in a `batch` module, and then import everything into its parent if the feature is enabled.

Maybe the cleanest way to do that would be to split `recognition/mod.rs` into `recognition/batch.rs` and `recognition/<normal, default, singlethread, someting like that...>`, then `mod.rs` would always import the normal recognizer but only import the batch recognizer if the feature is enabled. Then do a similar thing with the models, with a `models/` folder, a `models/mod.rs`, `models/batch.rs`, and `models/<normal, ... whatever you decided to name the recognizer file>`.

lightningpwr28 commented:
I added the batch modules in-file to allow for backwards compatibility. If we moved the non-batch to a different file, I think users would have to change the way they imported non-batch functions. This way is a little messier, but not *that* much messier.

Bear-03 commented on vosk/src/recognition/mod.rs:
```diff
@@ -277,3 +283,86 @@ impl Drop for Recognizer {
+
+        /// Gets the number of chunks that have yet to be processed
+        pub fn get_pending_chunks(&mut self) -> usize {
+            (unsafe { vosk_batch_recognizer_get_pending_chunks(self.0.as_ptr()) }) as usize
```
I don't think this should return `usize` as it'd be potentially wasting memory. The real number will only have the range of a `c_int` (`i32`) as per the underliying type in Vosk.

I suggest the following: correct me if I'm wrong, but I am pretty sure the chunk count will only be `>= 0` since it doesn't make sense to have "a negative of chunks left" and I checked the underlying code and it doesn't seem to use negative numbers for error codes. If that's the case, we could change the return type of this function to a `u16` to better describe the real range of the returned values.

lightningpwr28 replied:
Using a `u16` would probably be fine, but all the `CudaOnlinePipelineDynamicBatcher` stuff is pretty esoteric. 16 bits isn't that much memory to use just to be sure, but if you feel really strongly about it, I (or you) can change it to `u16`.

Bear-03 replied:
Nah it was just a suggestion. I have barely no background on GPU programming so I prefer to follow your advice. Plus in a way it's also better to just leave it as the original type since down the road it could prevent issues caused by unforseen values returned by vosk. This way is just simpler overall ;)

Bear-03 reviewed (changes requested):
I think the batch module part could be organized better but I have to think about it so I'll handle it myself before publishing the release that will include this.

Let me know what you think about the comment, and when that's changed I can proceed with the merge :P

Btw the clippy CI pipeline failed because of some CI issue I have to fix, (github is fetching the old vosk-sys crate, not the one you pushed) so don't worry about it.

Bear-03 reviewed (approved):
Everything looks good to me, merging ;)

Bear-03 commented:
Hey @lightningpwr28 I'm organizing and cleaning up PRS to prepare the crate for a new release and I was wondering, would you happen to have some examples of the JSON returned by `BatchRecognizer::front_result`? I would want that function to return a Rust type instead of a `serde_json` object so it is consistent with the result functions in `Recognizer`, but I cannot test this myself since I don't have a device with CUDA.
